### PR TITLE
Jerry Wu Take Home Assessment

### PR DESCRIPTION
**Pull request by nchannelmosfet**

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 repos:
   - repo: https://github.com/timothycrosley/isort
-    rev: 5.10.1
+    rev: 5.11.5
```
Comment (nchannelmosfet) [inline .pre-commit-config.yaml:3]: Upgraded the version to be compatible with M1 Mac

File: main.py [hunk 2]
```diff
@@ -18,12 +23,14 @@ async def get_users():
         statement = select(User)
         results = session.execute(statement)
         results = list(i[0] for i in results.all())
-    if len(results) == 0:
```
Comment (nchannelmosfet) [inline main.py:21]: Unnecessary length check. We can just return `result` even if it is an empty list.

File: main.py [hunk 3]
```diff
@@ -18,12 +23,14 @@ async def get_users():
...
-        return []
     return results
 
 
-@app.delete("/users/{user_id}", status_code=status.HTTP_204_NO_CONTENT)
+@app.delete(
+    "/users/{user_id}",
+    status_code=status.HTTP_204_NO_CONTENT,
+    responses={status.HTTP_404_NOT_FOUND: {"model": Message}},
```
Comment (nchannelmosfet) [inline main.py:32]: Added additional response models to communicate the possible response status in OpenAPI doc.

File: main.py [hunk 4]
```diff
@@ -41,3 +48,67 @@ async def create_user(user: User):
...
+    "/crops",
+    status_code=status.HTTP_201_CREATED,
+    response_model=Crop,
+    responses={status.HTTP_400_BAD_REQUEST: {"model": Message}},
+)
+async def create_crop(crop: Crop):
+    if not crop.tilled and crop.tillage_depth and crop.tillage_depth > 0:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Field is not tilled and should not have tillage_depth",
```
Comment (nchannelmosfet) [inline main.py:81]: I am not familiar with farming, but I will make an assumption that if a field is not tilled, it does not have tillage_depth. So, I added this validation.

File: main.py [hunk 5]
```diff
@@ -41,3 +48,67 @@ async def create_user(user: User):
...
+        result = session.execute(statement).first()
+    if not result:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND, detail=f"crop_id {crop_id} not found"
+        )
+    return result[0]
+
+
+@app.get("/crops", status_code=status.HTTP_200_OK, response_model=list[Crop])
+async def get_crops(limit: int = Query(10, le=100), offset: int = 0):
```
Comment (nchannelmosfet) [inline main.py:109]: Added `limit` and `offset` query parameters as a form pagination.

File: models.py [hunk 6]
```diff
@@ -1,8 +1,29 @@
...
+class CropType(str, Enum):
+    CORN: str = "corn"
+    WHEAT: str = "wheat"
+    BARLEY: str = "barkley"
+    HOPS: str = "hops"
 
 
 class User(SQLModel, table=True):
-    id: Optional[int] = Field(default=None, primary_key=True)
+    id: int | None = Field(default=None, primary_key=True)
```
Comment (nchannelmosfet) [inline models.py:14]: Out of habit, I will use Python3.10+ style for optional field.

File: models.py [hunk 7]
```diff
@@ -1,8 +1,29 @@
...
     name: str
+
+    crops: list["Crop"] = Relationship(back_populates="user")
+
+
+class Crop(SQLModel, table=True):
+    id: int | None = Field(default=None, primary_key=True)
+    year: int = Field(ge=1000, le=9999)
```
Comment (nchannelmosfet) [inline models.py:22]: Ensure year is 4 digits.

File: models.py [hunk 8]
```diff
@@ -1,8 +1,29 @@
-from typing import Optional
+from enum import Enum
 
-from sqlmodel import Field, SQLModel
+from sqlmodel import Field, Relationship, SQLModel
+
+
+class CropType(str, Enum):
```
Comment (nchannelmosfet) [inline models.py:6]: Use Enum to achieve constrained pick list.

File: test_main.py [hunk 9]
```diff
@@ -4,23 +4,83 @@
 from main import app
 
 
-def create_user(client) -> dict:
+def create_user(client: TestClient) -> dict:
```
Comment (nchannelmosfet) [inline test_main.py:7]: Added type annotation for convenience of intellisense.

File: test_main.py [hunk 10]
```diff
@@ -34,6 +94,34 @@ def test_crud_users():
...
+        # get 1 entry
+        response = get_crop(client, crop_id)
+        assert response["year"] == 2000
+        assert response["crop_type"] == "corn"
+
+        delete_crop(client, crop_id=crop_id)
+        response = get_all_crops(client)
+        assert len(response) == 0
+
+        delete_crop_404_error(client, crop_id=crop_id)
```
Comment (nchannelmosfet) [inline test_main.py:125]: Test negative paths

File: models.py [hunk 11]
```diff
@@ -1,8 +1,29 @@
...
+
+    crops: list["Crop"] = Relationship(back_populates="user")
+
+
+class Crop(SQLModel, table=True):
+    id: int | None = Field(default=None, primary_key=True)
+    year: int = Field(ge=1000, le=9999)
+    crop_type: CropType
+    tilled: bool
+    tillage_depth: float | None = Field(ge=0, lt=10)
```
Comment (nchannelmosfet) [inline models.py:25]: Constrain the range according to the requirement: 0 <= x < 10

File: models.py [hunk 12]
```diff
@@ -1,8 +1,29 @@
...
+    comments: str
+
+    user_id: int | None = Field(default=None, foreign_key="user.id")
```
Comment (nchannelmosfet) [inline models.py:28]: Foreign key connection with User model meaning user has crop.

**Comment by sigerber** [comment]
Hi Jerry, thank you for your submission!

I do apologise that our "README.MD" should **not** have asked you to raise a PR. We changed the way we administer our take-home assignments and this part of the instructions was left over from the old way. (Documentation is never correct, right? 🤦🏻‍♂️).

We have reviewed your submission but I will be going to close and attempt to remove this PR so your solution is not exposed to other candidates.

**Pull request closed without merging**